### ISSUE TITLE
Improving build setup

### ISSUE THREAD

**Issue by j-haj** [elastic/ml-cpp#457]
The current build setup process is quite involved, requiring the user to download specific libraries and make edits to specific files using editor-specific commands. Simplifying the process would lower the barrier to entry for new contributors and possibly reduce potential future build issues/errors.

It might be simpler to, at a minimum, fork these libraries and make the edits in the forked versions. This would allow contributors to simply clone the modified libraries rather than clone and modify locally. It would also standardize the changes.

Another approach would be to create a set of patches along with a build script that would acquire the appropriate dependencies, apply the necessary patches, and install.

I am happy to work on this if there is interest. I was going to submit a PR to remove `boost::array` this morning but can't test at the moment because I haven't gone through the build setup process.

**Comment by droberts195** [comment]
Which platform are you on?  If it's Mac or Linux then a simple thing you can do is install `docker`, then run `dev-tools/docker_test.sh linux` from the root of the ml-cpp repo.  That will copy your local changes into a Docker container and run the unit tests.  If you can get that to work then that's good enough to raise a PR, and the PR CI will find any Windows-specific problems.  (For Docker on Mac I changed the settings to: 3 CPUs, 4GB RAM and 2GB swap - with the default settings it ran out of memory while building.)

If you're on Windows, look at the "simple" version of the setup in `build-setup/windows_simple.md`.

> a build script that would acquire the appropriate dependencies, apply the necessary patches, and install

Have a look at `dev-tools/vagrant/linux/provision.sh`.  It was originally written to provision a Vagrant VM but is probably not very hard to adapt to be a more general provisioning script for any Linux machine.

> It might be simpler to, at a minimum, fork these libraries and make the edits in the forked versions

_Maybe_ we could do this with `log4cxx` as it is basically abandoned (and is also the component that requires the most edits).  But I don't like the idea of doing this for an actively developed repo like Boost for example.

**Comment by j-haj** [comment]
I am on Linux but can also test on Mac. The Linux Docker build using `docker_test.sh` seemed to work fine. If this is the preferred approach to building and testing, it may be worth calling this out first in the README (I am happy to submit a PR).

Regardless, I really do think having some automated way to apply all these changes is a good move. I will check out `dev-tools/vagrant/linux/provision.sh` and compare this with the `docker_test.sh` file.

**Comment by droberts195** [comment]
> If this is the preferred approach to building and testing

I don't think it is for people who spend a lot of time working on this repo.  `docker_test.sh` builds everything from scratch every time you run it.  It will get very frustrating for interactive development where you want to make a change, rebuild _just_ the files you touched and re-run tests.

For an interactive Docker environment, try this:

```
docker run --name my_ml_container -it docker.elastic.co/ml-dev/ml-linux-build:6 /bin/bash
```

The `6` in that line should be the value of `VERSION` in `dev-tools/docker/build_linux_build_image.sh` on the branch you're working on.

Inside the Docker container, run this:

```
cd
mkdir .ssh
scp dave@my_machine:.ssh/id_rsa* .ssh
git clone git@github.com:elastic/ml-cpp.git
cd ml-cpp
source set_env.sh
make -j `nproc`
make -j `nproc` test
exit
```

Obviously replace `my_machine` and `dave` in that with the name of your development machine and your username on your development machine.

Then to restart development from where you left off, only having to rebuild files that depend on your changes, use:

```
docker start -ia my_ml_container
```

The only problem with this is that within the container you're limited to text editors like `vim`.  (Unless you know a good way to use a fancy graphical editor within a Docker container...)

Otherwise, creating a script similar to `dev-tools/vagrant/linux/provision.sh` to install everything directly onto your development machine sounds like a good way to go.  I'm happy if you want to create such a script and raise a PR to add it into `build-setup`.

**Comment by j-haj** [comment]
Thanks! I will work on something similar to `dev-tools/vagrant/linux/provision.sh`
